pkg/security/scc/matcher.go

```diff
@@ -65,31 +66,113 @@ func ConstraintAppliesTo(constraint *kapi.SecurityContextConstraints, userInfo u
+		}
+	}
+	if requiresPreAllocatedSELinuxLevel(constraint) {
+		level, err := getPreallocatedLevel(namespace)
+		if err != nil {
+			return nil, fmt.Errorf("unable to find pre-allocated mcs annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+		}
+
+		// SELinuxOptions is a pointer, if we are resolving and it is already initialized
+		// we need to make a copy of it so we don't manipulate the store's cache.
+		if constraint.SELinuxContext.SELinuxOptions != nil {
+			seLinuxOptionsCopy := *constraint.SELinuxContext.SELinuxOptions
+			constraint.SELinuxContext.SELinuxOptions = &seLinuxOptionsCopy
+		} else {
+			constraint.SELinuxContext.SELinuxOptions = &kapi.SELinuxOptions{}
+		}
+		constraint.SELinuxContext.SELinuxOptions.Level = level
+	}
+	if requiresPreallocatedFSGroup(constraint) {
+		fsGroup, err := getPreallocatedFSGroup(namespace)
+		if err != nil {
+			return nil, fmt.Errorf("unable to find pre-allocated group annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+		}
+		constraint.FSGroup.Ranges = fsGroup
+	}
+	if requiresPreallocatedSupplementalGroups(constraint) {
+		supplementalGroups, err := getPreallocatedSupplementalGroups(namespace)
+		if err != nil {
+			return nil, fmt.Errorf("unable to find pre-allocated group annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+		}
+		constraint.SupplementalGroups.Ranges = supplementalGroups
+	}
+
+	// Create the provider
+	provider, err := kscc.NewSimpleProvider(constraint)
+	if err != nil {
+		namespaceName := ""
+		if namespace != nil {
+			namespaceName = namespace.Name
+		}
+		return nil, fmt.Errorf("error creating provider for SCC %s in namespace %s: %v", constraint.Name, namespaceName, err)
+	}
+	return provider, nil
+}
+
+// SetSecurityContext assigns the securityContext the annotations and the container
+// security contexts to the provided pod
+func SetSecurityContext(pod *kapi.Pod, psc *kapi.PodSecurityContext, annotations map[string]string, securityContexts []*kapi.SecurityContext) {
+	pod.Spec.SecurityContext = psc
+	pod.Annotations = annotations
+	for i := range pod.Spec.InitContainers {
+		pod.Spec.InitContainers[i].SecurityContext = securityContexts[i]
+	}
+	base := len(pod.Spec.InitContainers)
+	for i := range pod.Spec.Containers {
+		pod.Spec.Containers[i].SecurityContext = securityContexts[i+base]
+	}
+}
+
+// ResolvePodSecurityContext performs a non mutating check of the pod against a security provider.
+// All the containers must validate aginst the same provider.
+// It returns the PodSecurityContext the annotations and all containers SecurityContexts.
+func ResolvePodSecurityContext(provider kscc.SecurityContextConstraintsProvider, pod *kapi.Pod) (*kapi.PodSecurityContext, map[string]string, []*kapi.SecurityContext, field.ErrorList) {
```

Comment:
wdyt about splitting the returned SCs into init containers vs regular containers or a map of container name to SC (I don't think we allow name collisions)?  Doing the math in multiple functions to add a base to the index will probably be a future bug.